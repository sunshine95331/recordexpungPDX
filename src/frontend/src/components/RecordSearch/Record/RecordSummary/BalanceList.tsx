import React from "react";
import currencyFormat from "../../../../service/currency-format";

interface Props {
  displayTotal?: boolean;
  listItems: any[];
  title: string;
  totalBalance: number;
}

export default class BalanceList extends React.Component<Props> {
  render() {
    const { displayTotal, listItems, title, totalBalance } = this.props;

    return (
      <>
        <h3 className="fw7 bt b--light-gray pt2 mb3">{title}</h3>
        <ul className="mw5 list">
          {listItems.length > 0 ? listItems : "None"}
        </ul>
        {displayTotal && (
          <div className="mw5 bt b--light-gray pt2">
            <span className="fw6">Total</span>
            <span className="fr">{currencyFormat(totalBalance)}</span>
          </div>
        )}
      </>
    );
  }
}
