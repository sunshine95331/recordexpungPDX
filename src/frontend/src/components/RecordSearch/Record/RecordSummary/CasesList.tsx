export default function CasesList() {
  return <></>;
}
