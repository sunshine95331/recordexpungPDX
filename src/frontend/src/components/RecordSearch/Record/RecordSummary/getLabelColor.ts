export default function getElibilityColor(eligibility: string) {
  return (
    {
      "Eligible Now": "green",
      Ineligible: "red",
      "Needs More Analysis": "purple",
    }[eligibility] ?? "dark-blue"
  );
}
