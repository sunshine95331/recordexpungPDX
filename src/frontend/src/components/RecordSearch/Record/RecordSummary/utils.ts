import moment from "moment";

export function getEligibilityColor(eligibility: string) {
  return (
    {
      "Eligible Now": "green",
      Ineligible: "red",
      "Needs More Analysis": "purple",
    }[eligibility] ?? "dark-blue"
  );
}

export function getFinalEligibility(
  status: string,
  dateEligible: string,
  caseBalance: number
) {
  let finalStatus = status;

  if (dateEligible) {
    const eligibleDate = moment(dateEligible, "MMM D, YYYY");
    const isEligibleNow = moment().isSameOrAfter(eligibleDate);

    if (finalStatus === "Eligible") {
      if (isEligibleNow) {
        finalStatus += caseBalance > 0 ? " Now If Balance Paid" : " Now";
      } else {
        finalStatus += " " + dateEligible;
      }
    }
  }

  return finalStatus;
}
