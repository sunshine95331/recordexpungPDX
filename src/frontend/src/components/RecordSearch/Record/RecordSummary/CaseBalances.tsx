import React from "react";
import BalanceList from "./BalanceList";
import { CaseData } from "../types";
import currencyFormat from "../../../../service/currency-format";

interface Props {
  cases?: CaseData[];
  totalBalance: number;
}

const hasBalanceDue = (element: CaseData) => {
  return element.balance_due > 0;
};

export default class CaseBalances extends React.Component<Props> {
  render() {
    if (!this.props.cases) {
      return null;
    }

    const listItems = this.props.cases
      .filter(hasBalanceDue)
      .map((element: CaseData) => {
        return (
          <li className="mb2" key={element.case_number}>
            <a
              className="link underline hover-blue break-all"
              href={`#${element.case_number}`}
            >
              {element.case_number}
            </a>{" "}
            <span className="fr">{currencyFormat(element.balance_due)}</span>
          </li>
        );
      });

    return (
      <div className="mb4">
        <BalanceList
          listItems={listItems}
          title="Balance Due by Case"
          totalBalance={this.props.totalBalance}
        />
      </div>
    );
  }
}
