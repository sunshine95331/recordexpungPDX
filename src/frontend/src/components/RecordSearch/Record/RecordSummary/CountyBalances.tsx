import React from "react";
import BalanceList from "./BalanceList";
import { CountyBalanceData } from "../types";
import currencyFormat from "../../../../service/currency-format";

interface Props {
  balances: CountyBalanceData[];
  totalBalance: number;
}

export default class CountyBalances extends React.Component<Props> {
  render() {
    const listItems = this.props.balances.map(
      (pair: CountyBalanceData, i: number) => {
        return (
          <li className="mb2" key={i}>
            <span className="break-all">{pair.county_name} </span>{" "}
            <span className="fr">{currencyFormat(pair.balance)}</span>
          </li>
        );
      }
    );

    return (
      <BalanceList
        listItems={listItems}
        title="Balance Due by County"
        totalBalance={this.props.totalBalance}
      />
    );
  }
}
