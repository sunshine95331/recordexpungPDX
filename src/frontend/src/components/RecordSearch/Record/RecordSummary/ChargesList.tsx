import React from "react";
import { getEligibilityColor } from "./utils";

interface ChargeProps {
  id: string;
  name: string;
  hasBalance: boolean;
}

interface ChargesProps {
  caseWithBalanceHeader: string;
  charges: any[];
}

interface EligibilityGroupProps {
  eligibility: string;
  headerAndCharges: any[];
}

interface ChargeListProps {
  chargesGroupedByEligibilityAndCase: { [label: string]: any[] };
  totalCharges: number;
}

function Charge({ id, name, hasBalance }: ChargeProps) {
  return (
    <li className="f6 bb b--light-gray pv2">
      <a href={"#" + id} className={hasBalance ? "ml2" : "link hover-blue"}>
        {name}
      </a>
    </li>
  );
}

function Charges({ caseWithBalanceHeader, charges }: ChargesProps) {
  return (
    <>
      {caseWithBalanceHeader && (
        <li key={caseWithBalanceHeader} className="fw7 pt2">
          {caseWithBalanceHeader}
        </li>
      )}
      {charges.map(([id, name]: [string, string]) => {
        return (
          <Charge
            key={id}
            id={id}
            name={name}
            hasBalance={caseWithBalanceHeader !== ""}
          />
        );
      })}
    </>
  );
}

function EligibilityGroup({
  eligibility,
  headerAndCharges,
}: EligibilityGroupProps) {
  const labelColor = getEligibilityColor(eligibility);
  const numCharges = headerAndCharges.reduce(
    (acc: number, aCase: any) => acc + aCase[1].length,
    0
  );

  return (
    <div className="mb3">
      <div className={labelColor + " bb b--light-gray lh-copy pb1"}>
        <span className="fw7 mb2">{eligibility}</span>{" "}
        {numCharges > 0 && `(${numCharges})`}
        {eligibility === "Needs More Analysis" && (
          <p className="f6 fw5">
            These charges need clarification below before an accurate analysis
            can be determined
          </p>
        )}
      </div>

      <ul className="list">
        {headerAndCharges.map(([caseWithBalanceHeader, charges], idx) => (
          <Charges
            key={idx}
            caseWithBalanceHeader={caseWithBalanceHeader}
            charges={charges}
          />
        ))}
      </ul>
    </div>
  );
}

export default function ChargesList({
  chargesGroupedByEligibilityAndCase,
  totalCharges,
}: ChargeListProps) {
  return (
    <>
      <h3 className="bt b--light-gray pt2 mb3">
        <span className="fw7">Charges</span> ({totalCharges})
        <p className="f6 fw5 pt1">Excludes traffic violations</p>
      </h3>

      {Object.entries(chargesGroupedByEligibilityAndCase).map(
        ([eligibility, headerAndCharges]) => {
          return (
            <EligibilityGroup
              key={eligibility}
              eligibility={eligibility}
              headerAndCharges={headerAndCharges}
            />
          );
        }
      )}
    </>
  );
}
