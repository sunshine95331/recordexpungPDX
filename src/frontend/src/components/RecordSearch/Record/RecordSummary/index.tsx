import React from "react";
import CaseBalances from "./CaseBalances";
import ChargesList from "./ChargesList";
import CountyBalances from "./CountyBalances";
import { CaseData, RecordSummaryData } from "../types";
import { downloadPdf } from "../../../../redux/search/actions";

interface Props {
  cases?: CaseData[];
  summary: RecordSummaryData;
}

export default class RecordSummary extends React.Component<Props> {
  handleDownloadClick = () => {
    downloadPdf();
  };
  render() {
    const {
      total_charges,
      eligible_charges_by_date,
      county_balances,
      total_balance_due,
      total_cases,
    } = this.props.summary;

    return (
      <div className="bg-white shadow br3 mb3 ph3 pb3">
        <div className="flex justify-between">
          <h2 className="mv3 f5 fw7">Search Summary</h2>
          <button
            onClick={this.handleDownloadClick}
            className="ma2 ba bw1 b--light-gray mid-gray bg-white link hover-blue fw6 br3 pv1 ph2"
          >
            Download as PDF
          </button>
        </div>
        <div className="flex-ns flex-wrap">
          <ChargesList
            eligibleChargesByDate={eligible_charges_by_date}
            totalCases={total_cases}
            totalCharges={total_charges}
          />
          <div className="w-100 w-33-l ph3-l mb3">
            <CaseBalances
              cases={this.props.cases}
              totalBalance={total_balance_due}
            />
            {total_balance_due > 0 && (
              <CountyBalances balances={county_balances} />
            )}
          </div>
        </div>
      </div>
    );
  }
}
