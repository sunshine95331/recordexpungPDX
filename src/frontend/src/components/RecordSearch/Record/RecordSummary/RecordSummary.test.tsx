import React from "react";
import "@testing-library/jest-dom";
import { fireEvent, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { appRender } from "../../../../test/testHelpers";
import history from "../../../../service/history";
import RecordSummary from ".";

jest.mock("axios", () => {
  return {
    request: () => new Promise(() => {}),
  };
});

// snapshot tests for the initial state are accounted
// for in the RecordSearch tests
it("correctly renders with the complex fake record and CasesList view selected", () => {
  const { getByLabelText, asFragment } = appRender(
    <RecordSummary />,
    "complex"
  );

  fireEvent.click(getByLabelText("Cases"));
  expect(asFragment()).toMatchSnapshot();
});

describe("When rendered with the John Common demo data", () => {
  beforeEach(() => {
    appRender(<RecordSummary />, "common");
  });

  test("the generate paperwork button works", async () => {
    const user = userEvent.setup();
    const historySpy = jest.spyOn(history, "push").mockReturnValue();
    const generateButton = screen.getByRole("button", {
      name: /generate paperwork/i,
    });

    await user.click(generateButton);
    expect(historySpy).toHaveBeenCalledWith("/fill-expungement-forms");
  });

  test("the summary pdf button works", async () => {
    const user = userEvent.setup();
    const pdfButton = screen.getByRole("button", {
      name: /summary pdf/i,
    });

    await user.click(pdfButton);
    expect(pdfButton).toHaveClass("loading-btn");
  });
});

describe("When rendered with empty results", () => {
  beforeEach(() => {
    appRender(<RecordSummary />, "blank");
  });

  test("the generate paperwork button will display an error with a dismiss button", async () => {
    const user = userEvent.setup();
    const generateButton = screen.getByRole("button", {
      name: /generate paperwork/i,
    });
    const errorMessage = /must be eligible charges to generate/i;

    await user.click(generateButton);
    expect(screen.queryByText(errorMessage)).toBeInTheDocument();

    const dismissButton = screen.getByRole("button", {
      name: /close/i,
    });

    await user.click(dismissButton);
    expect(screen.queryByText(errorMessage)).not.toBeInTheDocument();
  });
});
