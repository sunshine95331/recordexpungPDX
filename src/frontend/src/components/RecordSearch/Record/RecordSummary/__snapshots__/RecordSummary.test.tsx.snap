// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correctly renders with the John Common demo data 1`] = `
<div
  className="bg-white shadow br3 mb3 ph3 pb3"
>
  <div
    className="flex flex-wrap justify-end mb1"
  >
    <h2
      className="f5 fw7 mv3 mr-auto"
    >
      Search Summary
    </h2>
    <button
      className="ma2 nowrap mid-gray link hover-blue fw6 br3 pv1 ph2"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="fas fa-bolt pr2"
      />
      Generate Paperwork
    </button>
    <button
      className="ma2 nowrap mid-gray link hover-blue fw6 br3 pv1 ph2"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="fas fa-download pr2"
      />
      Summary PDF
    </button>
  </div>
  <div
    className="flex-ns flex-wrap"
  >
    <div
      className="w-100 w-two-thirds-l"
    >
      <h3
        className="bt b--light-gray pt2 mb3"
      >
        <span
          className="fw7"
        >
          Cases
        </span>
         (
        5
        )
      </h3>
      <h3
        className="bt b--light-gray pt2 mb3"
      >
        <span
          className="fw7"
        >
          Charges
        </span>
         (
        5
        )
        <p
          className="f6 fw5 pt1"
        >
          Excludes traffic violations
        </p>
      </h3>
      <div
        className="mb3"
      >
        <div
          className="dark-blue bb b--light-gray lh-copy pb1"
        >
          <span
            className="fw7 mb2"
          >
            Eligible Nov 28, 2025
          </span>
           
          (1)
        </div>
        <ul
          className="list"
        >
          <li
            className="f6 bb b--light-gray pv2"
          >
            <a
              className="link hover-blue"
              href="#100000-1"
            >
              Aggravated Theft in the First Degree (CONVICTED) Charged Jul 16, 2016
            </a>
          </li>
        </ul>
      </div>
      <div
        className="mb3"
      >
        <div
          className="green bb b--light-gray lh-copy pb1"
        >
          <span
            className="fw7 mb2"
          >
            Eligible Now
          </span>
           
          (3)
        </div>
        <ul
          className="list"
        >
          <li
            className="f6 bb b--light-gray pv2"
          >
            <a
              className="link hover-blue"
              href="#200000-1"
            >
              Obstruction of search warrant (DISMISSED) Charged Jul 16, 2019
            </a>
          </li>
          <li
            className="f6 bb b--light-gray pv2"
          >
            <a
              className="link hover-blue"
              href="#110000-1"
            >
              Theft in the Second Degree (CONVICTED) Charged Apr 2, 2015
            </a>
          </li>
          <li
            className="f6 bb b--light-gray pv2"
          >
            <a
              className="link hover-blue"
              href="#120000-1"
            >
              Poss under oz Marijuana (CONVICTED) Charged Apr 2, 2014
            </a>
          </li>
        </ul>
      </div>
      <div
        className="mb3"
      >
        <div
          className="purple bb b--light-gray lh-copy pb1"
        >
          <span
            className="fw7 mb2"
          >
            Needs More Analysis
          </span>
           
          (1)
          <p
            className="f6 fw5"
          >
            These charges need clarification below before an accurate analysis can be determined
          </p>
        </div>
        <ul
          className="list"
        >
          <li
            className="f6 bb b--light-gray pv2"
          >
            <a
              className="link hover-blue"
              href="#210000-1"
            >
              Poss Controlled Sub (CONVICTED) Charged Sep 23, 2018
            </a>
          </li>
        </ul>
      </div>
    </div>
    <div
      className="w-100 w-33-l ph3-l mb3"
    >
      <div
        className="mb4"
      >
        <h3
          className="fw7 bt b--light-gray pt2 mb3"
        >
          Balance due by county
           
          <a
            className=" gray link hover-blue underline"
            href="/manual#paybalances"
            onClick={[Function]}
          >
            <i
              aria-hidden="true"
              className="fas fa-question-circle link hover-dark-blue"
            />
            <span
              className="visually-hidden"
            >
              Learn more about balances
            </span>
          </a>
        </h3>
        <ul
          className="mw5 list"
        >
          None
        </ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`correctly renders with the Multiple Charges demo data 1`] = `
<div
  className="bg-white shadow br3 mb3 ph3 pb3"
>
  <div
    className="flex flex-wrap justify-end mb1"
  >
    <h2
      className="f5 fw7 mv3 mr-auto"
    >
      Search Summary
    </h2>
    <button
      className="ma2 nowrap mid-gray link hover-blue fw6 br3 pv1 ph2"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="fas fa-bolt pr2"
      />
      Generate Paperwork
    </button>
    <button
      className="ma2 nowrap mid-gray link hover-blue fw6 br3 pv1 ph2"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="fas fa-download pr2"
      />
      Summary PDF
    </button>
  </div>
  <div
    className="flex-ns flex-wrap"
  >
    <div
      className="w-100 w-two-thirds-l"
    >
      <h3
        className="bt b--light-gray pt2 mb3"
      >
        <span
          className="fw7"
        >
          Cases
        </span>
         (
        3
        )
      </h3>
      <h3
        className="bt b--light-gray pt2 mb3"
      >
        <span
          className="fw7"
        >
          Charges
        </span>
         (
        4
        )
        <p
          className="f6 fw5 pt1"
        >
          Excludes traffic violations
        </p>
      </h3>
      <div
        className="mb3"
      >
        <div
          className="green bb b--light-gray lh-copy pb1"
        >
          <span
            className="fw7 mb2"
          >
            Eligible Now
          </span>
           
          (1)
        </div>
        <ul
          className="list"
        >
          <li
            className="f6 bb b--light-gray pv2"
          >
            <a
              className="link hover-blue"
              href="#110000-1"
            >
              Theft in the Third Degree (DISMISSED) Charged Jan 21, 2022
            </a>
          </li>
        </ul>
      </div>
      <div
        className="mb3"
      >
        <div
          className="dark-blue bb b--light-gray lh-copy pb1"
        >
          <span
            className="fw7 mb2"
          >
            Eligible Now If Balance Paid
          </span>
           
          (2)
        </div>
        <ul
          className="list"
        >
          <li
            className="fw7 pt2"
          >
            Baker 100000 â€“ $1000.0
          </li>
          <li
            className="f6 bb b--light-gray pv2"
          >
            <a
              className="ml2"
              href="#100000-1"
            >
              Disorderly Conduct in the First Degree (CONVICTED) Charged Jan 21, 2019
            </a>
          </li>
          <li
            className="f6 bb b--light-gray pv2"
          >
            <a
              className="ml2"
              href="#100000-2"
            >
              Disorderly Conduct in the Second Degree (DISMISSED) Charged Jan 21, 2019
            </a>
          </li>
        </ul>
      </div>
    </div>
    <div
      className="w-100 w-33-l ph3-l mb3"
    >
      <div
        className="mb4"
      >
        <h3
          className="fw7 bt b--light-gray pt2 mb3"
        >
          Balance due by county
           
          <a
            className=" gray link hover-blue underline"
            href="/manual#paybalances"
            onClick={[Function]}
          >
            <i
              aria-hidden="true"
              className="fas fa-question-circle link hover-dark-blue"
            />
            <span
              className="visually-hidden"
            >
              Learn more about balances
            </span>
          </a>
        </h3>
        <ul
          className="mw5 list"
        >
          <div>
            <h4
              className="fw6 mb2"
            >
              Baker
            </h4>
            <ul
              className="mw5 list mb3"
            >
              <li
                className="mb2"
              >
                <a
                  className="link hover-blue"
                  href="/#100000"
                  onClick={[Function]}
                >
                  100000
                </a>
                 
                <span
                  className="fr"
                >
                  $1,000.00
                </span>
              </li>
            </ul>
          </div>
          <div
            className="mw5 bt b--light-gray pt2"
          >
            <span
              className="fw6"
            >
              Total
            </span>
            <span
              className="fr"
            >
              $1,000.00
            </span>
          </div>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
